dali/core/mm/cuda_vm_resource_test.cc

```diff
@@ -0,0 +1,305 @@
+    auto &region = res.va_regions_.back();
+    ASSERT_EQ(region.num_blocks(), b1 + b2 + b3);
+    ComparePart(region, 0, va1);
+    ComparePart(region, b1 + b2, va3);
+    for (int i = b1; i < b1 + b2; i++) {
+      EXPECT_FALSE(region.mapped[i]);
+      EXPECT_FALSE(region.mapped[i]);
+    }
+  }
+
+  std::mt19937_64 rng_{12345};
+};
+
+TEST_F(VMResourceTest, BasicTest) {
+  if (!cuvm::IsSupported())
+    GTEST_SKIP() << "CUDA Virtual Memory Management not supported on this platform";
+  this->TestAlloc();
+}
+
+TEST_F(VMResourceTest, RegionMerge) {
+  if (!cuvm::IsSupported())
+    GTEST_SKIP() << "CUDA Virtual Memory Management not supported on this platform";
+  this->TestRegionMerge();
+}
+
+TEST_F(VMResourceTest, RegionExtendAfter) {
+  if (!cuvm::IsSupported())
+    GTEST_SKIP() << "CUDA Virtual Memory Management not supported on this platform";
+  this->TestRegionExtendAfter();
+}
+
+TEST_F(VMResourceTest, RegionExtendBefore) {
+  if (!cuvm::IsSupported())
+    GTEST_SKIP() << "CUDA Virtual Memory Management not supported on this platform";
+  this->TestRegionExtendBefore();
+}
+
+std::string format_size(size_t bytes) {
+  std::stringstream ss;
+  print(ss, bytes, " bytes");
+  if (bytes > (1LLU << 34)) {
+    print(ss, " (", bytes >> 30, " GiB)");
+  } else if (bytes > (1LLU << 24)) {
+    print(ss, " (", bytes >> 20, " MiB)");
+  } else if (bytes > (1LLU << 14)) {
+    print(ss, " (", bytes >> 10, " KiB)");
+  }
+  return ss.str();
+}
+
+using perfclock = std::chrono::high_resolution_clock;
+
+template <typename Out = double, typename R, typename P>
+inline Out microseconds(std::chrono::duration<R, P> d) {
+  return std::chrono::duration_cast<std::chrono::duration<Out, std::micro>>(d).count();
+}
+
+TEST_F(VMResourceTest, RandomAllocations) {
+  if (!cuvm::IsSupported())
+    GTEST_SKIP() << "CUDA Virtual Memory Management not supported on this platform";
+
+  cuda_vm_resource pool(-1, 4 << 20);  // use small 4 MiB blocks
+  std::mt19937_64 rng(12345);
+  std::bernoulli_distribution is_free(0.5);
+  std::uniform_int_distribution<int> align_dist(0, 8);  // alignment anywhere from 1B to 256B
+  std::poisson_distribution<int> size_dist(4 << 20);  // average at 4 MiB
+  struct allocation {
+    void *ptr;
+    size_t size, alignment;
+  };
+  std::vector<allocation> allocs;
+
+  int num_iter = 100000;
+
+  std::map<uintptr_t, size_t> allocated;
+
+  perfclock::duration alloc_time = {};
+  perfclock::duration dealloc_time = {};
+
+  for (int i = 0; i < num_iter; i++) {
+    if (is_free(rng) && !allocs.empty()) {
+      auto idx = rng() % allocs.size();
+      allocation a = allocs[idx];
+      auto t0 = perfclock::now();
+      pool.deallocate(a.ptr, a.size, a.alignment);
+      auto t1 = perfclock::now();
+      dealloc_time += t1 - t0;
+      std::swap(allocs[idx], allocs.back());
+      allocs.pop_back();
+    } else {
+      allocation a;
+      a.size = std::max(1, std::min(size_dist(rng), 256 << 20));  // Limit to 256 MiB
```

Comment:
Doesn't `size_dist` has smaller range than `256 << 20`?

Reply:
It's a Poisson distribution - it's not bounded - it's just that the probability of high values is very small, but they can occur.